components/service/sync-autofill/src/main/java/mozilla/components/service/sync/autofill/AutofillCreditCardsAddressesStorage.kt

```diff
@@ -24,34 +30,96 @@ const val AUTOFILL_DB_NAME = "autofill.sqlite"
  * library.
  */
 class AutofillCreditCardsAddressesStorage(
-    context: Context
-) : CreditCardsAddressesStorage, AutoCloseable {
+    context: Context,
+    securePrefs: Lazy<SecureAbove22Preferences>
+) : CreditCardsAddressesStorage, SyncableStore, AutoCloseable {
+    private val logger = Logger("AutofillCCAddressesStorage")
 
     private val coroutineContext by lazy { Dispatchers.IO }
 
+    val crypto by lazy { AutofillCrypto(context, securePrefs.value) }
+
     private val conn by lazy {
         AutofillStorageConnection.init(dbPath = context.getDatabasePath(AUTOFILL_DB_NAME).absolutePath)
         AutofillStorageConnection
     }
 
-    override suspend fun addCreditCard(creditCardFields: UpdatableCreditCardFields): CreditCard =
-        withContext(coroutineContext) {
-            conn.getStorage().addCreditCard(creditCardFields.into()).into()
+    private fun newKeyHandler(reason: KeyGenerationReason) {
+        when (reason) {
+            KeyGenerationReason.New -> logger.info("CC key generated")
+            // At this point, we need A-S API to recover: https://github.com/mozilla/application-services/issues/4015
+            KeyGenerationReason.Lost -> logger.warn("CC key lost, new one generated")
+            KeyGenerationReason.Corrupt -> logger.warn("CC key was corrupted, new one generated")
+            KeyGenerationReason.AbnormalState -> logger.warn("CC key lost due to storage malfunction, new one generated")
         }
-
-    override suspend fun getCreditCard(guid: String): CreditCard = withContext(coroutineContext) {
-        conn.getStorage().getCreditCard(guid).into()
     }
 
-    override suspend fun getAllCreditCards(): List<CreditCard> = withContext(coroutineContext) {
-        conn.getStorage().getAllCreditCards().map { it.into() }
+    override suspend fun addCreditCard(
+        creditCardFields: NewCreditCardFields
+    ): CreditCard = withContext(coroutineContext) {
+        // If we need to recover from key failure, do so now.
+        // We'll add the credit card with a new key below.
+        val key = crypto.key()
+        if (key.recoveryNeeded()) {
+            newKeyHandler(key.wasGenerated!!)
+        }
+
+        val updatableCreditCardFields = UpdatableCreditCardFields(
+            billingName = creditCardFields.billingName,
+            cardNumber = crypto.encrypt(key, creditCardFields.plaintextCardNumber),
```

Comment:
Once that recovery function is in place, we probably want to special case an empty plain text - I assume that encrypting empty plain text will still give you some ciphertext and that will avoid us re-filling the number from sync should we have it.  IOW, if plaintext is empty we should try and keep the ciphertext empty

Reply:
I think special-casing makes sense, but let's deal with that once we're adding actual recovery calls in here.